flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -74,6 +77,36 @@
 		}
 	}
 
+	/**
+	 * Verifies if the provided pattern can possibly generate empty match. Example of patterns that can possibly
+	 * generate empty matches are: A*, A?, A* B? etc.
+	 *
+	 * @param pattern pattern to check
+	 * @return true if empty match could potentially match the pattern, false otherwise
+	 */
+	public static boolean canProduceEmptyMatches(final Pattern<?, ?> pattern) {
+		NFAFactoryCompiler<?> compiler = new NFAFactoryCompiler<>(checkNotNull(pattern));
+		compiler.compileFactory();
+		State<?> startState = compiler.getStates().stream().filter(State::isStart).findFirst().get();
+
+		final Stack<State<?>> statesToCheck = new Stack<>();
+		statesToCheck.push(startState);
+		while (!statesToCheck.isEmpty()) {
+			final State<?> currentState = statesToCheck.pop();
+			for (StateTransition<?> transition : currentState.getStateTransitions()) {
+				if (transition.getAction() == StateTransitionAction.PROCEED) {
+					if (transition.getTargetState().isFinal()) {
+						return true;
+					} else {
+						statesToCheck.push(transition.getTargetState());
```

Comment:
Can we loop here forever if `PROCEED` takes us to already visited node? Shouldn't we check for already visited states?

Reply:
Theoretically yes, practically compiler won't produce such graph as of now. Do you think I should be defensive though?

Reply:
It would be better to at least throw `IllegalStateException` instead of looping forever in case of some bug or if someone will refactor/add new feature to the compiler.